File: 0135-candy/0135-candy.ts
function candy(ratings: number[]): number {
    const n = ratings.length;

    if (!ratings || n === 0) {
        return 0;
    }

    let total = 1;
    let i = 1;

    while (i < n) {
        if (ratings[i] === ratings[i - 1]) {
            total++;
            i++;
            continue;
        }

        let peak = 1;
        while (i < n && ratings[i] > ratings[i - 1]) {
            total += ++peak;
            i++;
        }

        let down = 1;
        while (i < n && ratings[i] < ratings[i - 1]) {
            total += down++;
            i++;
        }

        if (peak < down) {
            total += down - peak;
        }
    }

    return total;
}

File: 0135-candy/0135-candy.java
class Solution {
    public int candy(int[] ratings) {
        int n = ratings.length;

        if(ratings == null || n == 0){
            return 0;
        }

        int[] candies = new int[n];
        candies[0] = 1;

        for(int i = 1; i < n; i++){
            candies[i] = (ratings[i] > ratings[i-1]) ? candies[i-1]+1 : 1;
        }

        for(int i = n-2; i >= 0; i--){
            candies[i] = (ratings[i] > ratings[i+1]) ? 
            Math.max(candies[i],candies[i+1]+1) : Math.max(1,candies[i]);
        }

        int total = 0;
        for(int cndy: candies){
            total += cndy;
        }

        return total;
    }
}